File: device/input-bridge.js
#!/usr/bin/env node
/**
 * Hardware Input Bridge for Spotify Car Thing
 * Reads from /dev/input/event* and forwards to server via WebSocket
 */

const { spawn } = require('child_process');
const WebSocket = require('ws');

// Event device mapping (from /proc/bus/input/devices)
const DEVICES = {
  BUTTONS: '/dev/input/event0',  // gpio-keys (preset buttons, back button)
  DIAL: '/dev/input/event1'       // rotary@0 (dial rotation)
};

// Key codes from linux/input-event-codes.h
const KEY_CODES = {
  KEY_BACK: 158,
  KEY_ENTER: 28,
  KEY_LEFT: 105,   // Dial counter-clockwise
  KEY_RIGHT: 106,  // Dial clockwise
  BTN_0: 256,      // Preset button 1
  BTN_1: 257,      // Preset button 2
  BTN_2: 258,      // Preset button 3
  BTN_3: 259       // Preset button 4
};

// Event types
const EV_KEY = 0x01;
const EV_REL = 0x02;
const REL_DIAL = 0x07;

class InputBridge {
  constructor(serverUrl = 'ws://127.0.0.1:3000') {
    this.serverUrl = serverUrl;
    this.ws = null;
    this.reconnectDelay = 1000;
    this.maxReconnectDelay = 30000;
  }

  connect() {
    console.log('[Input Bridge] Connecting to server:', this.serverUrl);
    
    this.ws = new WebSocket(`${this.serverUrl}/ws/device?id=input-bridge`);
    
    this.ws.on('open', () => {
      console.log('[Input Bridge] Connected to server');
      this.reconnectDelay = 1000;
    });
    
    this.ws.on('close', () => {
      console.log('[Input Bridge] Disconnected from server, reconnecting in', this.reconnectDelay, 'ms');
      setTimeout(() => this.connect(), this.reconnectDelay);
      this.reconnectDelay = Math.min(this.reconnectDelay * 2, this.maxReconnectDelay);
    });
    
    this.ws.on('error', (err) => {
      console.error('[Input Bridge] WebSocket error:', err.message);
    });
  }

  sendInput(keyCode, isPressed) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      const message = {
        type: 'input',
        keyCode,
        isPressed
      };
      this.ws.send(JSON.stringify(message));
      console.log('[Input Bridge] Sent:', message);
    }
  }

  startButtonMonitor() {
    console.log('[Input Bridge] Starting button monitor:', DEVICES.BUTTONS);
    
    const getevent = spawn('getevent', ['-l', DEVICES.BUTTONS], {
      stdio: ['ignore', 'pipe', 'pipe']
    });

    getevent.stdout.on('data', (data) => {
      const lines = data.toString().split('\n');
      for (const line of lines) {
        this.parseButtonEvent(line);
      }
    });

    getevent.on('error', (err) => {
      console.error('[Input Bridge] Button monitor error:', err);
    });

    getevent.on('exit', (code) => {
      console.log('[Input Bridge] Button monitor exited with code:', code);
      setTimeout(() => this.startButtonMonitor(), 1000);
    });
  }

  startDialMonitor() {
    console.log('[Input Bridge] Starting dial monitor:', DEVICES.DIAL);
    
    const getevent = spawn('getevent', ['-l', DEVICES.DIAL], {
      stdio: ['ignore', 'pipe', 'pipe']
    });

    getevent.stdout.on('data', (data) => {
      const lines = data.toString().split('\n');
      for (const line of lines) {
        this.parseDialEvent(line);
      }
    });

    getevent.on('error', (err) => {
      console.error('[Input Bridge] Dial monitor error:', err);
    });

    getevent.on('exit', (code) => {
      console.log('[Input Bridge] Dial monitor exited with code:', code);
      setTimeout(() => this.startDialMonitor(), 1000);
    });
  }

  parseButtonEvent(line) {
    // Example: EV_KEY       KEY_BACK             DOWN
    // Example: EV_KEY       BTN_0                UP
    const match = line.match(/EV_KEY\s+(\S+)\s+(DOWN|UP)/);
    if (!match) return;

    const [, keyName, state] = match;
    const isPressed = state === 'DOWN';

    // Map key names to codes
    let keyCode = null;
    switch (keyName) {
      case 'KEY_BACK':
        keyCode = KEY_CODES.KEY_BACK;
        break;
      case 'BTN_0':
        keyCode = KEY_CODES.BTN_0;
        break;
      case 'BTN_1':
        keyCode = KEY_CODES.BTN_1;
        break;
      case 'BTN_2':
        keyCode = KEY_CODES.BTN_2;
        break;
      case 'BTN_3':
        keyCode = KEY_CODES.BTN_3;
        break;
      case 'KEY_ENTER':
        keyCode = KEY_CODES.KEY_ENTER;
        break;
    }

    if (keyCode !== null) {
      this.sendInput(keyCode, isPressed);
    }
  }

  parseDialEvent(line) {
    // Example: EV_REL       REL_DIAL             00000001
    // Example: EV_REL       REL_DIAL             ffffffff
    const match = line.match(/EV_REL\s+REL_DIAL\s+([0-9a-f]{8})/i);
    if (!match) return;

    const valueHex = match[1];
    const value = parseInt(valueHex, 16);
    
    // Convert to signed 32-bit integer
    const signedValue = value > 0x7fffffff ? value - 0x100000000 : value;

    // Positive = clockwise = RIGHT, Negative = counter-clockwise = LEFT
    if (signedValue > 0) {
      // Clockwise rotation
      this.sendInput(KEY_CODES.KEY_RIGHT, true);
      setTimeout(() => this.sendInput(KEY_CODES.KEY_RIGHT, false), 50);
    } else if (signedValue < 0) {
      // Counter-clockwise rotation
      this.sendInput(KEY_CODES.KEY_LEFT, true);
      setTimeout(() => this.sendInput(KEY_CODES.KEY_LEFT, false), 50);
    }
  }

  start() {
    console.log('[Input Bridge] Starting...');
    this.connect();
    this.startButtonMonitor();
    this.startDialMonitor();
  }
}

// Start the bridge
const bridge = new InputBridge();
bridge.start();

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('[Input Bridge] Shutting down...');
  process.exit(0);
});

File: device/input-bridge.py
#!/usr/bin/env python3
"""
Hardware Input Bridge for Spotify Car Thing
Reads from /dev/input/event* and sends to server via HTTP
Uses direct event reading (no getevent required)
"""

import struct
import json
import time
from threading import Thread
from urllib import request
from urllib.error import URLError

# Event device mapping
DEVICE_BUTTONS = '/dev/input/event0'  # gpio-keys (buttons)
DEVICE_DIAL = '/dev/input/event1'      # rotary encoder

# Key codes (actual values from hardware)
KEY_ESC = 1        # Back button
KEY_1 = 2          # Preset 1
KEY_2 = 3          # Preset 2
KEY_3 = 4          # Preset 3
KEY_4 = 5          # Preset 4
KEY_ENTER = 28     # Dial click

# Virtual key codes to send to server
KEY_BACK = 158
KEY_LEFT = 105
KEY_RIGHT = 106
BTN_0 = 256
BTN_1 = 257
BTN_2 = 258
BTN_3 = 259

# Event types
EV_KEY = 0x01
EV_REL = 0x02
REL_DIAL = 0x06

SERVER_URL = 'http://127.0.0.1:3000'

# Input event struct format: timeval (2 longs), type (short), code (short), value (int)
# On 32-bit ARM: LL (8 bytes) HH (4 bytes) i (4 bytes) = 16 bytes
EVENT_FORMAT = 'llHHi'
EVENT_SIZE = struct.calcsize(EVENT_FORMAT)

def send_input(key_code, is_pressed):
    """Send input event to server via HTTP POST"""
    try:
        data = json.dumps({
            'deviceId': 'input-bridge',
            'keyCode': key_code,
            'isPressed': is_pressed
        }).encode('utf-8')
        
        req = request.Request(
            f'{SERVER_URL}/api/input',
            data=data,
            headers={'Content-Type': 'application/json'},
            method='POST'
        )
        
        with request.urlopen(req, timeout=1) as response:
            pass
        
        print(f'[Input Bridge] Sent: keyCode={key_code}, isPressed={is_pressed}')
    except Exception as e:
        print(f'[Input Bridge] Error sending: {e}')

def monitor_buttons():
    """Monitor button events from event0"""
    print(f'[Input Bridge] Starting button monitor: {DEVICE_BUTTONS}')
    
    # Map hardware codes to virtual codes we send to server
    key_map = {
        KEY_ESC: KEY_BACK,      # Back button -> virtual BACK
        KEY_ENTER: KEY_ENTER,   # Dial click -> virtual ENTER  
        KEY_1: BTN_0,           # Preset 1
        KEY_2: BTN_1,           # Preset 2
        KEY_3: BTN_2,           # Preset 3
        KEY_4: BTN_3            # Preset 4
    }
    
    while True:
        try:
            with open(DEVICE_BUTTONS, 'rb') as f:
                while True:
                    event_data = f.read(EVENT_SIZE)
                    if len(event_data) < EVENT_SIZE:
                        break
                    
                    tv_sec, tv_usec, ev_type, code, value = struct.unpack(EVENT_FORMAT, event_data)
                    
                    print(f'[DEBUG] BUTTON event: type={ev_type:02x}, code={code:02x}, value={value}')
                    
                    if ev_type == EV_KEY and code in key_map:
                        # value: 0 = UP, 1 = DOWN, 2 = REPEAT
                        if value == 1:  # DOWN
                            send_input(key_map[code], True)
                        elif value == 0:  # UP
                            send_input(key_map[code], False)
        
        except Exception as e:
            print(f'[Input Bridge] Button monitor error: {e}')
        
        time.sleep(1)  # Retry delay

def monitor_dial():
    """Monitor dial rotation events from event1"""
    print(f'[Input Bridge] Starting dial monitor: {DEVICE_DIAL}')
    
    while True:
        try:
            with open(DEVICE_DIAL, 'rb') as f:
                while True:
                    event_data = f.read(EVENT_SIZE)
                    if len(event_data) < EVENT_SIZE:
                        break
                    
                    tv_sec, tv_usec, ev_type, code, value = struct.unpack(EVENT_FORMAT, event_data)
                    
                    print(f'[DEBUG] DIAL event: type={ev_type:02x}, code={code:02x}, value={value}')
                    
                    if ev_type == EV_REL and code == REL_DIAL:
                        # Convert to signed if needed (though car thing sends +1/-1 already)
                        if value > 0x7fffffff:
                            value = value - 0x100000000
                        
                        # Positive = clockwise = RIGHT, Negative = counter-clockwise = LEFT
                        if value > 0:
                            send_input(KEY_RIGHT, True)
                            time.sleep(0.01)
                            send_input(KEY_RIGHT, False)
                        elif value < 0:
                            send_input(KEY_LEFT, True)
                            time.sleep(0.01)
                            send_input(KEY_LEFT, False)
        
        except Exception as e:
            print(f'[Input Bridge] Dial monitor error: {e}')
        
        time.sleep(1)  # Retry delay

if __name__ == '__main__':
    print('[Input Bridge] Starting hardware input bridge...')
    
    # Start both monitors in separate threads
    button_thread = Thread(target=monitor_buttons, daemon=True)
    dial_thread = Thread(target=monitor_dial, daemon=True)
    
    button_thread.start()
    dial_thread.start()
    
    # Keep main thread alive
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print('[Input Bridge] Shutting down...')
